### PR TITLE
Addition of Flink Metrics in Atleast Once Implementation.

### PR DESCRIPTION
**Pull request by prashastia**
View this [doc](https://docs.google.com/document/d/17YPHMKROHzi1pUEZKUkBTzXg2D5-d1vaiXErJZv3YTU/edit?usp=sharing) for implementation details of this PR.

/gcbrun

### CODE REVIEW

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/BigQueryDefaultSink.java [hunk 1]
```diff
@@ -38,6 +38,11 @@ class BigQueryDefaultSink extends BigQueryBaseSink {
     public SinkWriter createWriter(InitContext context) {
         checkParallelism(context.getNumberOfParallelSubtasks());
         return new BigQueryDefaultWriter(
-                context.getSubtaskId(), tablePath, connectOptions, schemaProvider, serializer);
+                context.getSubtaskId(),
+                tablePath,
+                connectOptions,
+                schemaProvider,
+                serializer,
+                context);
```
Comment (jayehwhyehentee) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/BigQueryDefaultSink.java:46]: If we're passing the whole context, then no need to send `context.getSubtaskId()` separately.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java [hunk 2]
```diff
@@ -56,15 +57,15 @@
  * validated lazily.
  *
  * <p>Serializer's "init" method is called in the writer's constructor because the resulting {@link
- * Descriptor} is not serializable and cannot be propagated to machines hosting writer instances.
- * Hence, this derivation of descriptors must be performed during writer initialization.
+ * com.google.protobuf.Descriptors.Descriptor} is not serializable and cannot be propagated to
```
Comment (jayehwhyehentee) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:60]: Why are we replacing with the fully qualified name?

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java [hunk 3]
```diff
@@ -56,15 +57,15 @@
...
+ * machines hosting writer instances. Hence, this derivation of descriptors must be performed during
+ * writer initialization.
  *
  * @param <IN> Type of records to be written to BigQuery.
  */
 abstract class BaseWriter<IN> implements SinkWriter<IN> {
 
-    protected final Logger logger = LoggerFactory.getLogger(getClass());
-
+    protected final Logger logger = LoggerFactory.getLogger(BaseWriter.class);
```
Comment (jayehwhyehentee) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:68]: Why is this needed?
https://stackoverflow.com/questions/12160540/java-logging-with-abstract-classes

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java [hunk 4]
```diff
@@ -90,6 +91,10 @@ abstract class BaseWriter<IN> implements SinkWriter<IN> {
     // the records in a stream get committed to the table. Hence, records written to BigQuery
     // table is equal to this "totalRecordsWritten" only upon checkpoint completion.
     long totalRecordsWritten;
+    // Counters for metric reporting
+    Counter successfullyAppendedRecordsCounter;
+    Counter successfullyAppendedRecordsSinceChkptCounter;
+    Counter numRecordsInSinceChkptCounter;
```
Comment (jayehwhyehentee) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:97]: Please remove the `Counter` suffix from these attribute names. It is redundant.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java [hunk 5]
```diff
@@ -116,6 +121,11 @@ public void flush(boolean endOfInput) {
         }
         logger.info("Validating all pending append responses in subtask {}", subtaskId);
         validateAppendResponses(true);
+        // .flush() is called at checkpoint, resetting the counters after all tasks are done.
```
Comment (jayehwhyehentee) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:124]: Instead of `.flush()`, just `Writer's flush` is more readable.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java [hunk 6]
```diff
@@ -134,6 +144,15 @@ public void close() {
...
         }
+        if (numRecordsInSinceChkptCounter != null) {
+            numRecordsInSinceChkptCounter = null;
+        }
+        if (successfullyAppendedRecordsSinceChkptCounter != null) {
+            successfullyAppendedRecordsSinceChkptCounter = null;
+        }
+        if (successfullyAppendedRecordsCounter != null) {
+            successfullyAppendedRecordsCounter = null;
+        }
```
Comment (jayehwhyehentee) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:155]: Is there no better way of closing a counter?
If not, then do we need to explicitly set it to null?

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryBufferedWriter.java [hunk 7]
```diff
@@ -98,7 +99,8 @@ public BigQueryBufferedWriter(
             long totalRecordsWritten,
             BigQueryConnectOptions connectOptions,
             BigQuerySchemaProvider schemaProvider,
-            BigQueryProtoSerializer serializer) {
+            BigQueryProtoSerializer serializer,
+            Sink.InitContext context) {
```
Comment (jayehwhyehentee) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryBufferedWriter.java:103]: Can we import `InitContext` directly?

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/BigQueryDefaultSink.java (hunk 1 above)
Comment (prashastia) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/BigQueryDefaultSink.java]: Fixed.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java (hunk 2 above)
Comment (prashastia) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java]: imported the missing dependency.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java (hunk 4 above)
Comment (prashastia) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java]: Done.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java (hunk 5 above)
Comment (prashastia) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java]: Fixed.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryBufferedWriter.java (hunk 7 above)
Comment (prashastia) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryBufferedWriter.java]: Done.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java (hunk 3 above)
Comment (prashastia) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:68]: Uniformity, should we change to `getClass()` for every class that uses a logger?

Comment (jayehwhyehentee) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java]: `getClass()` should be used only when the class type is not known an compile time. BaseWriter in itself is an abstract class. The logs defined in it will be executed by its implementations (ALO and EXO writers) at runtime.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java (hunk 6 above)
Comment (prashastia) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java]: Removed the explicit assignment.

**Review by jayehwhyehentee — APPROVED** [review]
LGTM. Please ensure presubmits are passing

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryBufferedWriter.java [hunk 8]
```diff
@@ -108,6 +109,15 @@ public BigQueryBufferedWriter(
         this.totalRecordsWritten = totalRecordsWritten;
         writeStreamCreationThrottler = new WriteStreamCreationThrottler(subtaskId);
         appendRequestRowCount = 0L;
+        // Initialize the metric counters.
+        successfullyAppendedRecords = context.metricGroup().counter("successfullyAppendedRecords");
+        // Update the metrics to values saved at checkpoint.
+        successfullyAppendedRecords.inc(totalRecordsWritten);
+        context.metricGroup().getIOMetricGroup().getNumRecordsInCounter().inc(totalRecordsSeen);
+        // ..SinceChkpt Counters restart at 0.
```
Comment (prashastia) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryBufferedWriter.java:117]: NumberOfRecordsSeenByWriter

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryBufferedWriter.java [hunk 9]
```diff
@@ -108,6 +109,15 @@ public BigQueryBufferedWriter(
         this.totalRecordsWritten = totalRecordsWritten;
         writeStreamCreationThrottler = new WriteStreamCreationThrottler(subtaskId);
         appendRequestRowCount = 0L;
+        // Initialize the metric counters.
+        successfullyAppendedRecords = context.metricGroup().counter("successfullyAppendedRecords");
```
Comment (prashastia) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryBufferedWriter.java:113]: NumberOfRecordsWritten

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java [hunk 10]
```diff
@@ -18,6 +18,7 @@
 
 import org.apache.flink.annotation.Internal;
 import org.apache.flink.api.connector.sink2.SinkWriter;
+import org.apache.flink.metrics.Counter;
```
Comment (prashastia) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:21]: Common repetition to be moved to common class

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java [hunk 11]
```diff
@@ -64,7 +66,6 @@
 abstract class BaseWriter<IN> implements SinkWriter<IN> {
 
     protected final Logger logger = LoggerFactory.getLogger(getClass());
-
     // Multiply 0.95 to keep a buffer from exceeding payload limits.
     private static final long MAX_APPEND_REQUEST_BYTES =
```
Comment (prashastia) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:70]: initalizeMetrics() abstract in base class and then specific implementation in their specific classes.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryBufferedWriter.java [hunk 12]
```diff
@@ -17,6 +17,7 @@
 package com.google.cloud.flink.bigquery.sink.writer;
 
 import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.connector.sink2.Sink.InitContext;
```
Comment (prashastia) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryBufferedWriter.java:20]: Should not be visibly defined in BufferedWriter.

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java [hunk 13]
```diff
@@ -116,6 +121,11 @@ public void flush(boolean endOfInput) {
         }
         logger.info("Validating all pending append responses in subtask {}", subtaskId);
         validateAppendResponses(true);
+        // Writer's flush() is called at checkpoint, resetting the counters after all tasks are
+        // done. Set to 0.
+        this.numRecordsSinceCheckpoint.dec(this.numRecordsSinceCheckpoint.getCount());
```
Comment (prashastia) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:126]: Remove this as EXO does not follow this
